File: ICPC/2015 Divisionals/lights/submissions/lights-max.cpp
#include <bits/stdc++.h>

using namespace std;

int main() {
	int N, D;
	cin >> N >> D;
	vector<vector<int>> lights(N, vector<int>(4));
	for (auto &v : lights) for (int &e : v) cin >> e;
	for (auto &l : lights) {
		int len = l[2]+l[3], t = l[0]-l[1], rem = t%len;
		if (rem > l[2] || t < 0) {
			cout << "NO" << endl;
			return 0;
		}
	}
	cout << "YES" << endl;
}

File: ICPC/2015 Divisionals/lights/submissions/LightsMax.java
import java.util.*;

class LightsMax {
	static class Light {
		public int x;
		public int a;
		public int b;
		public int c;
		public Light() {
		}
	}
	public static void main(String[] args) {
		Scanner sc = new Scanner(System.in);
		int n = sc.nextInt();
		int d = sc.nextInt();
		Light[] lights = new Light[n];
		for (int i = 0; i < n; ++i) {
			lights[i] = new Light();
			lights[i].x = sc.nextInt();
			lights[i].a = sc.nextInt();
			lights[i].b = sc.nextInt();
			lights[i].c = sc.nextInt();
		}
		boolean g = true;
		for (int i = 0; i < n; ++i) {
			int len = lights[i].b+lights[i].c;
			int t = lights[i].x-lights[i].a;
			int rem = t%len;
			if (rem > lights[i].b || t < 0) {
				g = false;
				break;
			}
		}
		System.out.println(g ? "YES" : "NO");

	}
}
